### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [2]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [5]:
players = purchase_data['Purchase ID'].count()
players

780

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [15]:
ave_price = purchase_data['Price'].mean()
revenue = purchase_data['Price'].sum()
print(ave_price)
print(revenue)

3.050987179487176
2379.77


In [11]:
item_num = len(purchase_data['Item ID'].unique())
item_num

183

In [16]:
Summary_Table = pd.DataFrame({'Unique Items': [183],
                 'Average Price': [3.05],
                 'Total Purchase': [780],
                 'Revenue': [2378.77]})
Summary_Table

Unique Items  Average Price  Total Purchase  Revenue
0           183           3.05             780  2378.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [20]:
gender_count = purchase_data['Gender'].value_counts()
gender_count

Male                     652
Female                   113
Other / Non-Disclosed     15
Name: Gender, dtype: int64

In [21]:
percentage = gender_count/780*100
percentage

Male                     83.589744
Female                   14.487179
Other / Non-Disclosed     1.923077
Name: Gender, dtype: float64

In [51]:
summary_table2 = pd.DataFrame({'Gender': ['Male', 'Female', 'Non-Dis'],
                                'Total Count': [652, 113, 15],
                              'Percentage': [83.58, 14.48, 1.92],})
summary_table2.set_index('Gender')

Total Count  Percentage
Gender                          
Male             652       83.58
Female           113       14.48
Non-Dis           15        1.92


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [28]:
male_analysis = purchase_data.loc[purchase_data["Gender"] == "Male", :]
male_analysis.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [34]:
male_price = male_analysis['Price'].mean()
male_count = male_analysis['SN'].count()
total_m = male_analysis['Price'].sum()
per_person = total_m/ male_count

print(male_price)
print(male_count)
print(total_m)
print(per_person)

3.0178527607361953
652
1967.64
3.0178527607361967


In [47]:
female_analysis = purchase_data.loc[purchase_data["Gender"] == "Female", :]
non_analysis = purchase_data.loc[purchase_data["Gender"] == "Other / Non-Disclosed", :]

In [48]:
female_price = female_analysis['Price'].mean()
female_count = female_analysis['SN'].count()
total_f = female_analysis['Price'].sum()
per_personf = total_f/ female_count
non_price = non_analysis['Price'].mean()
non_count = non_analysis['SN'].count()
total_n = non_analysis['Price'].sum()
per_personn = total_n/ non_count

In [49]:
print(female_price)
print(female_count)
print(total_f)
print(per_personf)

3.203008849557519
113
361.94
3.203008849557522


In [50]:
print(non_price)
print(non_count)
print(total_n)
print(per_personn)

3.3460000000000005
15
50.19
3.3459999999999996


In [52]:
summary_table3 = pd.DataFrame({'Gender': ['Male', 'Female', 'Non-Dis'],
                                'Total Count': [652, 113, 15],
                              'Ave Purcahse Price': [3.01, 3.20, 3.34],
                              'Total Purchase Value': [1967.64, 361.94, 50.19],
                              'Avg Total Purchase Per Person': [3.01, 3.20, 3.34]})
summary_table3.set_index('Gender')

Total Count  Ave Purcahse Price  Total Purchase Value  \
Gender                                                           
Male             652                3.01               1967.64   
Female           113                3.20                361.94   
Non-Dis           15                3.34                 50.19   

         Avg Total Purchase Per Person  
Gender                                  
Male                              3.01  
Female                            3.20  
Non-Dis                           3.34

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [53]:
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [57]:
bins = [0, 10, 20, 30, 40, 50]
age_label = ["0 - 10", "11-20", "21-30", "31-40", "41-59"]
pd.cut(purchase_data["Age"], bins, labels=age_label).head()

0    11-20
1    31-40
2    21-30
3    21-30
4    21-30
Name: Age, dtype: category
Categories (5, object): [0 - 10 < 11-20 < 21-30 < 31-40 < 41-59]

In [61]:
purchase_data['Age Bracket'] = pd.cut(purchase_data["Age"], bins, labels=age_label)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Bracket  
0  Extraction, Quickblade Of Trembling Hands   3.53       11-20  
1                          Frenzied Scimitar   1.56       31-40  
2                               Final Critic   4.88       21-30  
3                                Blindscythe   3.27       21-30  
4                                       Fury   1.44       21-30

In [69]:
purchase_count = purchase_data['Age Bracket'].value_counts()

In [68]:
groupby_age = purchase_data.groupby(['Age Bracket'])
purchase_total = groupby_age['Price'].sum()

In [73]:
purchase_mean = groupby_age['Price'].mean()

In [70]:
per_age = purchase_total/purchase_count
per_age.head()

0 - 10    3.405000
11-20     3.063622
21-30     2.992687
31-40     3.153647
41-59     3.075714
dtype: float64

In [75]:
summary_table3= pd.DataFrame({"Purchase Count": purchase_count, "Purchase Total": purchase_total, 
                              "Average Price": purchase_mean, "Purchase per Age": per_age})
summary_table3.head()

Purchase Count  Purchase Total  Average Price  Purchase per Age
Age Bracket                                                                 
0 - 10                   32          108.96       3.405000          3.405000
11-20                   254          778.16       3.063622          3.063622
21-30                   402         1203.06       2.992687          2.992687
31-40                    85          268.06       3.153647          3.153647
41-59                     7           21.53       3.075714          3.075714

In [80]:
Percentage= purchase_count/780*100
summary_table3['Percentage'] = Percentage
summary_table3.head()

Purchase Count  Purchase Total  Average Price  Purchase per Age  \
Age Bracket                                                                    
0 - 10                   32          108.96       3.405000          3.405000   
11-20                   254          778.16       3.063622          3.063622   
21-30                   402         1203.06       2.992687          2.992687   
31-40                    85          268.06       3.153647          3.153647   
41-59                     7           21.53       3.075714          3.075714   

             Percentage  
Age Bracket              
0 - 10         4.102564  
11-20         32.564103  
21-30         51.538462  
31-40         10.897436  
41-59          0.897436

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

Purchase Count Average Purchase Price Total Purchase Value  \
10-14             28                  $2.96               $82.78   
15-19            136                  $3.04              $412.89   
20-24            365                  $3.05            $1,114.06   
25-29            101                  $2.90              $293.00   
30-34             73                  $2.93              $214.00   
35-39             41                  $3.60              $147.67   
40+               13                  $2.94               $38.24   
<10               23                  $3.35               $77.13   

      Avg Total Purchase per Person  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19  
<10                           $4.54

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [90]:
groupby_sn = purchase_data.groupby(['SN'])
total_buy = groupby_sn['Price'].sum()
sn_buy = total_buy.sort_values(ascending=False).head()
sn_buy

SN
Lisosia93      18.96
Idastidru52    15.45
Chamjask73     13.83
Iral74         13.62
Iskadarya95    13.10
Name: Price, dtype: float64

In [88]:
sn_count = groupby_sn['Price'].count()
sn_count.sort_values(ascending=False).head(10)

SN
Lisosia93       5
Iral74          4
Idastidru52     4
Aina42          3
Iri67           3
Chamjask73      3
Iskadarya95     3
Phyali88        3
Chanastnya43    3
Silaera56       3
Name: Price, dtype: int64

In [93]:
summary_table4 = pd.DataFrame({
                              "Total Purchase": sn_buy,
                              "Count": [5, 4, 3, 4, 3],
                              "Ave Purchase Price": [18.96/4, 15.45/4, 13.83/3, 13.62/4, 13.10/3]})
summary_table4

Total Purchase  Count  Ave Purchase Price
SN                                                    
Lisosia93             18.96      5            4.740000
Idastidru52           15.45      4            3.862500
Chamjask73            13.83      3            4.610000
Iral74                13.62      4            3.405000
Iskadarya95           13.10      3            4.366667

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [95]:
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Bracket  Percentage  
0  Extraction, Quickblade Of Trembling Hands   3.53       11-20         NaN  
1                          Frenzied Scimitar   1.56       31-40         NaN  
2                               Final Critic   4.88       21-30         NaN  
3                                Blindscythe   3.27       21-30         NaN  
4                                       Fury   1.44       21-30         NaN

In [115]:
item = purchase_data.loc[:, ['Item ID', 'Item Name', 'Price']]
item.head()

Item ID                                  Item Name  Price
0      108  Extraction, Quickblade Of Trembling Hands   3.53
1      143                          Frenzied Scimitar   1.56
2       92                               Final Critic   4.88
3      100                                Blindscythe   3.27
4      131                                       Fury   1.44

In [123]:
item_table = item.groupby(['Item ID', 'Item Name'])
item_total = item_table['Price'].sum()
item_count = item_table['Price'].count()
item_price = item_table['Price'].mean()
summary_table5 = pd.DataFrame({"Count": item_count, 
                              "Total": item_total,
                              'Price': item_price})
summary_table5.head()

Count  Total  Price
Item ID Item Name                              
0       Splinter                4   5.12   1.28
1       Crucifer                3   9.78   3.26
2       Verdict                 6  14.88   2.48
3       Phantomlight            6  14.94   2.49
4       Bloodlord's Fetish      5   8.50   1.70

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [127]:
summary_table5.sort_values(['Total'], axis= 0, ascending=False).head()

Count  Total  Price
Item ID Item Name                                                        
178     Oathbreaker, Last Hope of the Breaking Storm     12  50.76   4.23
82      Nirvana                                           9  44.10   4.90
145     Fiery Glass Crusader                              9  41.22   4.58
92      Final Critic                                      8  39.04   4.88
103     Singed Scalpel                                    8  34.80   4.35

Purchase Count  \
Item ID Item Name                                                     
178     Oathbreaker, Last Hope of the Breaking Storm             12   
82      Nirvana                                                   9   
145     Fiery Glass Crusader                                      9   
92      Final Critic                                              8   
103     Singed Scalpel                                            8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

three observations: item ID: 178 is most popular item
                                21-30 age bracket buys most games
                                lastly, the older the age the least likely to spend more on a game